src/compiler/parser.ts

```diff
@@ -2143,16 +2166,31 @@ module ts {
             }
         }
 
+        function setContextParameters(explicitFlags: ParserContextFlags = ParserContextFlags.None, ambientFlags: ParserContextFlags = ParserContextFlags.None) {
+            var savedFlags = contextFlags;

```

Comment:
I'll add a comment, but the gist is:

`setContextParameters` atomically returns the value for `contextFlags` at the time the function was called, and updates `contextFlags` in the following fashion:
* Any parser-generated flags not in the set of explicit and ambient flags are cleared from `contextFlags`
* Explicit flags are set on `contextFlags`.

For example:
```
contextFlags; // Yield | DisallowIn

var saveFlags = setContextParameters(/*explicitFlags* .Await, /*ambientFlags*/ .DisallowIn);

saveFlags; // Yield | DisallowIn
contextFlags; // Await | DisallowIn 
```